FREEWARE/v40/xpuzzles/src/xpyraminx/xpyraminx.man
.\" X-BASED PYRAMINX(tm)
.\"
.\" xpyraminx.man
.\"
.\" ##
.\"
.\" Copyright (c) 1994 - 95	David Albert Bagley
.\"
.\"                   All Rights Reserved
.\"
.\" Permission to use, copy, modify, and distribute this software and
.\" its documentation for any purpose and without fee is hereby granted,
.\" provided that the above copyright notice appear in all copies and
.\" that both that copyright notice and this permission notice appear in
.\" supporting documentation, and that the name of the author not be
.\" used in advertising or publicity pertaining to distribution of the
.\" software without specific, written prior permission.
.\"
.\" This program is distributed in the hope that it will be "playable",
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.TH XPYRAMINX 6 "16 May 1995" "V4.10"
.SH NAME
xpyraminx \- Pyramid X widget
.SH SYNOPSIS
.B /usr/games/xpyraminx
[-display <display>] [-geometry <geometry>] [-{size <int> | sticky}]
[-[no]orient] [-mono] [-{mode <int> | both}] [-{background | bg} <color>]
[-{foreground | fg} <color>] [-face{0|1|2|3} <color>]
.SH DESCRIPTION
.LP
Use "\fBmouse-left\fP" to move ccw.
.LP
Use "\fBmouse-center\fP", "\fBP\fP", or "\fBp\fP" to toggle the practice
mode (in practice mode the record should be "practice").  One must double
click on "\fBmouse-center\fP" if the puzzle is being worked on.
.LP
"\fBmouse-right\fP", "\fBR\fP", or "\fBr\fP", to randomize (this must be
done first to set a new record).  One must double click on
"\fBmouse-right\fP" if the puzzle is being worked on.
.LP
"\fBI\fP" or "\fBi\fP" keys "increase" the number of tetras.
.LP
"\fBD\fP" or "\fBd\fP" keys "decrease" the number of tetras.
.LP
"\fBO\fP" or "\fBo\fP" keys toggle the orient mode.
.LP
"\fB2\fP", "\fB3\fP", "\fBB\fP", or "\fBb\fP" keys change modes to
Period 2, Period 3, or Both.
.LP
"\fBY\fP" or "\fBy\fP" toggles sticky mode (increase/decrease is
disabled here).
.LP
"\fBS\fP" or "\fBs\fP" keys reserved for the auto-solver (unimplemented).
.LP
"\fBU\fP" or "\fBu\fP" keys to undo move.
.LP
"\fBE\fP" or "\fBe\fP" keys to enter a saved puzzle.
.LP
"\fBW\fP" or "\fBw\fP" keys to write or save a puzzle.
.LP
"\fBQ\fP", "\fBq\fP", or "\fBCTRL-C\fP" keys kill program.
.LP
Use the key pad or "R" keys for other ways to move other than CCW.
.LP
Use the control key and the left mouse button, keypad, or arrow keys to
move the whole tetrahedron.
.SH OPTIONS
.TP 8
.B \-display \fIhost\fP:\fIdpy\fP
This option specifies the X server to contact.
.TP 8
.B \-geometry {+|\-}\fIX\fP{+|\-}\fIY\fP
This option sets the initial position of the pyraminx window (resource
name "\fIgeometry\fP").
.TP 8
.B \-{background | bg} \fIcolor\fP
This option specifies the background of the pyraminx window (resource name
"\fIbackground\fP").
.TP 8
.B \-{foreground | fg} \fIcolor\fP
This option specifies the foreground of the pyraminx window (resource name
"\fIforeground\fP").
.TP 8
.B \-size <\fIint\fP>
This option allows you to change the number of tetras on a edge
(resource name "\fIsize\fP").
.TP 8
.B \-[no]orient
This option allows you to access the orient mode (resource name
"\fIorient\fP").
.TP 8
.B \-mode <\fIint\fP>
This option allows you to set the turning mode (resource name
"\fImode\fP").
.TP 8
.B \-both
This option allows you to set the turning mode to both period 2 and
period 3 (resource name "\fImode\fP" set at 4).
.TP 8
.B \-sticky
This option allows you to set the sticky mode (resource name
"\fIsticky\fP").
.TP 8
.B \-mono
This option allows you to access the mono mode on a color monitor
(resource name "\fImono\fP").
.TP 8
.B \-face{0|1|2|3} <\fIcolor\fP>
This option allows you to change the color of a face (resource name
"\fIfaceColorN\fP"). In mono-mode, color is represented as the first letter
of the color name. The faces are ordered top to bottom and left to right.
If you has two colors that begin with the same letter you should have one
in uppercase and one in lowercase to distinguish them in mono-mode. You can
change the colors of the faces to make a stupid pyraminx (i.e. all White or
in mono-mode all "W"). Unfortunately, it will not normally say its solved
when its randomized. This would be cheating.
.SH SEE ALSO
.LP
X(1), xoct(6), xrubik(6), xskewb(6), xdino(6), xmball(6), xmlink(6),
xcubes(6), xtriangles(6), xhexagons(6), xabacus(1) 
.SH COPYRIGHTS
.LP
\*R Copyright 1994-95, David Albert Bagley
.SH BUG REPORTS AND PROGRAM UPDATES
.LP
Send bugs (or their reports, or fixes) to the author
.RS
David Albert Bagley,	<\fIbagleyd@source.asset.com\fP>
.RE
.LP
The most updated source could be found on \fIftp.x.org\fP under
\fI/contrib/games/puzzles\fP.


FREEWARE/v30/xpuzzles/src/xpyraminx/xpyraminx.c
/*
# X-BASED PYRAMINX(tm)
#
#  xpyraminx.c
#
###
#
#  Copyright (c) 1994 - 95	David Albert Bagley, bagleyd@source.asset.com
#
#                   All Rights Reserved
#
#  Permission to use, copy, modify, and distribute this software and
#  its documentation for any purpose and without fee is hereby granted,
#  provided that the above copyright notice appear in all copies and
#  that both that copyright notice and this permission notice appear in
#  supporting documentation, and that the name of the author not be
#  used in advertising or publicity pertaining to distribution of the
#  software without specific, written prior permission.
#
#  This program is distributed in the hope that it will be "playable",
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
*/

/*
  Version 4: 94/05/31 Xt
  Version 3: 93/04/01 Motif
  Version 2: 92/01/29 XView
  Version 1: 91/03/19 SunView
*/

#include <stdlib.h>
#include <stdio.h>
#ifdef VMS
#include <unixlib.h>
#define getlogin cuserid
#else
#ifndef apollo
#include <unistd.h>
#endif
#endif
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Shell.h>
#include <X11/cursorfont.h>
#include "Pyraminx.h"
#include "pyraminx.xbm"

#ifndef SCOREFILE
#define SCOREFILE "/usr/games/lib/pyraminx.scores"
#endif
 
/* The following 6 are in PyraminxP.h also */
#define MINTETRAS 1
#define MAXTETRAS 7
#define PERIOD2 2
#define PERIOD3 3
#define BOTH 4
#define MAXMODES 3

#define MAXRECORD 32767
#define MAXPROGNAME 80
#define MAXNAME 256

static void initialize();
static void moves_text();

static void print_record();
static int handle_solved();
static void print_state();
static void read_records();
static void write_records();
 
static Arg arg[5];
static int pyraminx_record[MAXMODES][2][MAXTETRAS - MINTETRAS + 2];
static int moves_dsp = 0;
static char prog_dsp[64] = "xpyraminx";
static char record_dsp[16] = "INF";
static char message_dsp[128] = "Randomize to start";
static char title_dsp[256] = "";
static int oldsize;

static void usage()
{
  (void) fprintf(stderr, "usage: xpyraminx\n");
  (void) fprintf(stderr,
    "\t[-geometry [{width}][x{height}][{+-}{xoff}[{+-}{yoff}]]]\n");
  (void) fprintf(stderr,
    "\t[-display [{host}]:[{vs}]][-fg {color}] [-bg {color}]\n");
  (void) fprintf(stderr,
    "\t[-{size {int} | sticky}] [-[no]orient] [-mono]\n");
  (void) fprintf(stderr,
    "\t[-{mode {int} | both}] [-face{0|1|2|3} {color}]\n");
  exit(1);
}

static XrmOptionDescRec options[] = {
  {"-fg",		"*pyraminx.Foreground",	XrmoptionSepArg,	NULL},
  {"-bd",		"*pyraminx.Foreground",	XrmoptionSepArg,	NULL},
  {"-bg",		"*Background",		XrmoptionSepArg,	NULL},
  {"-foreground",	"*pyraminx.Foreground",	XrmoptionSepArg,	NULL},
  {"-background",	"*Background",		XrmoptionSepArg,	NULL},
  {"-bordercolor",	"*pyraminx.Foreground",	XrmoptionSepArg,	NULL},
  {"-size",		"*pyraminx.size",	XrmoptionSepArg,	NULL},
  {"-sticky",		"*pyraminx.sticky",	XrmoptionNoArg,		"FALSE"},
  {"-orient",		"*pyraminx.orient",	XrmoptionNoArg,		"TRUE"},
  {"-noorient",		"*pyraminx.orient",	XrmoptionNoArg,		"FALSE"},
  {"-mode",		"*pyraminx.mode",	XrmoptionSepArg,	NULL},
  {"-both",		"*pyraminx.mode",	XrmoptionNoArg,		"4"},
  {"-mono",		"*pyraminx.mono",	XrmoptionNoArg,		"TRUE"},
  {"-face0",		"*pyraminx.faceColor0",	XrmoptionSepArg,	NULL},
  {"-face1",		"*pyraminx.faceColor1",	XrmoptionSepArg,	NULL},
  {"-face2",		"*pyraminx.faceColor2",	XrmoptionSepArg,	NULL},
  {"-face3",		"*pyraminx.faceColor3",	XrmoptionSepArg,	NULL}
};

int main(argc, argv)
  int argc;
  char *argv[];
{
  Widget toplevel, pyraminx; 

  toplevel = XtInitialize(argv[0], "Pyraminx",
    options, XtNumber(options), &argc, argv);
  if (argc != 1)
    usage();

  XtSetArg(arg[0], XtNiconPixmap,
    XCreateBitmapFromData(XtDisplay(toplevel),
      RootWindowOfScreen(XtScreen(toplevel)),
      (char *) pyraminx_bits, pyraminx_width, pyraminx_height));
  XtSetValues(toplevel, arg, 1);
  pyraminx = XtCreateManagedWidget("pyraminx", pyraminxWidgetClass, toplevel,
    NULL, 0);
  XtAddCallback(pyraminx, XtNselectCallback, moves_text, NULL);
  initialize(pyraminx);
  XtRealizeWidget(toplevel);
  XGrabButton(XtDisplay(pyraminx), AnyButton, AnyModifier, XtWindow(pyraminx),
    TRUE, ButtonPressMask | ButtonMotionMask | ButtonReleaseMask,
    GrabModeAsync, GrabModeAsync, XtWindow(pyraminx),
    XCreateFontCursor(XtDisplay(pyraminx), XC_crosshair));
  XtMainLoop();

#ifdef VMS
  return 1;
#else
  return 0;
#endif
}

static void initialize(w)
  Widget w;
{
  int size, mode;
  Boolean orient, sticky;

  XtVaSetValues(w,
    XtNpractice, FALSE,
    XtNstart, FALSE,
    NULL);
  XtVaGetValues(w,
    XtNsize, &size,
    XtNmode, &mode,
    XtNorient, &orient,
    XtNsticky, &sticky,
    NULL);
  read_records();
  print_record(size, mode, orient, sticky, record_dsp);
  oldsize = size;
  print_state(XtParent(w), prog_dsp, mode, size, sticky, moves_dsp,
    record_dsp, message_dsp);
}

static void moves_text(w, client_data, call_data)
  Widget w;
  caddr_t client_data;
  pyraminxCallbackStruct *call_data;
{
  int size, mode;
  Boolean orient, practice, sticky;

  XtVaGetValues(w,
    XtNsize, &size,
    XtNorient, &orient,
    XtNmode, &mode,
    XtNorient, &orient,
    XtNsticky, &sticky,
    XtNpractice, &practice,
    NULL);
  (void) strcpy(message_dsp, "");
  switch (call_data->reason) {
    case PYRAMINX_RESTORE:
      if (practice)
        (void) strcpy(record_dsp, "practice");
      moves_dsp = 0;
      break;
    case PYRAMINX_RESET:
      moves_dsp = 0;
      break;
    case PYRAMINX_IGNORE:
      (void) strcpy(message_dsp, "Randomize to start");
      break;
    case PYRAMINX_MOVED:
      moves_dsp++;
#ifdef DEBUG
      if (moves_dsp > 256)
        exit(1);
#endif
      XtSetArg(arg[0], XtNstart, TRUE);
      XtSetValues(w, arg, 1);
      break;
    case PYRAMINX_CONTROL:
      return;
    case PYRAMINX_SOLVED:
      if (practice)
        moves_dsp = 0;
      else { 
        if (handle_solved(moves_dsp, size, mode, orient, sticky))
          (void) sprintf(message_dsp, "Congratulations %s!!", getenv("USER"));
        else
          (void) strcpy(message_dsp, "Solved!");
      }
      XtSetArg(arg[0], XtNstart, FALSE);
      XtSetValues(w, arg, 1);
      break;
    case PYRAMINX_PRACTICE:
      moves_dsp = 0;
      practice = !practice;
      if (practice)
        (void) strcpy(record_dsp, "practice");
      else {
        (void) strcpy(message_dsp, "Randomize to start");
        print_record(size, mode, orient, sticky, record_dsp);
      }
      XtSetArg(arg[0], XtNpractice, practice);
      XtSetArg(arg[1], XtNstart, FALSE);
      XtSetValues(w, arg, 2);
      break;
    case PYRAMINX_RANDOMIZE:
      moves_dsp = 0;
      XtSetArg(arg[0], XtNpractice, FALSE);
      XtSetArg(arg[1], XtNstart, FALSE);
      XtSetValues(w, arg, 2);
      break; 
    case PYRAMINX_DEC:
      if (!sticky) {
        moves_dsp = 0;
        size--;
        oldsize = size;
        print_record(size, mode, orient, sticky, record_dsp);
        XtSetArg(arg[0], XtNsize, size);
        XtSetValues(w, arg, 1);
      }
      break;
    case PYRAMINX_ORIENT:
      moves_dsp = 0;
      orient = !orient;
      print_record(size, mode, orient, sticky, record_dsp);
      XtSetArg(arg[0], XtNorient, orient);
      XtSetValues(w, arg, 1);
      break;
    case PYRAMINX_INC:
      if (!sticky) {
        moves_dsp = 0;
        size++;
        oldsize = size;
        print_record(size, mode, orient, sticky, record_dsp);
        XtSetArg(arg[0], XtNsize, size);
        XtSetValues(w, arg, 1);
      }
      break;
    case PYRAMINX_PERIOD2:
    case PYRAMINX_PERIOD3:
    case PYRAMINX_BOTH:
      moves_dsp = 0;
      mode = call_data->reason - PYRAMINX_PERIOD2 + PERIOD2;
      print_record(size, mode, orient, sticky, record_dsp);
      XtSetArg(arg[0], XtNmode, mode);
      XtSetValues(w, arg, 1);
      break;
    case PYRAMINX_STICKY:
      moves_dsp = 0;
      sticky = !sticky;
      if (sticky)
        size = 4;
      else
        size = oldsize;
      print_record(size, mode, orient, sticky, record_dsp);
      XtSetArg(arg[0], XtNsticky, sticky);
      XtSetArg(arg[1], XtNsize, size);
      XtSetValues(w, arg, 2);
      break;
    case PYRAMINX_COMPUTED:
      XtSetArg(arg[0], XtNstart, FALSE);
      XtSetValues(w, arg, 1);
      break;
    case PYRAMINX_UNDO:
      moves_dsp--;
      XtSetArg(arg[0], XtNstart, TRUE);
      XtSetValues(w, arg, 1);
      break;
  }
  print_state(XtParent(w), prog_dsp, mode, size, sticky, moves_dsp,
    record_dsp, message_dsp);
}

static void print_record(size, mode, orient, sticky, record)
  int size, mode;
  Boolean orient, sticky;
  char *record;
{
  int i = mode - PERIOD2;
  int j = (orient) ? 1 : 0;  
  int k = (sticky) ? MAXTETRAS - MINTETRAS + 1 : size - MINTETRAS;
 
  if (pyraminx_record[i][j][k] >= MAXRECORD)
    (void) strcpy(record, "NEVER");
  else
    (void) sprintf(record, "%d", pyraminx_record[i][j][k]);
}

static int handle_solved(counter, size, mode, orient, sticky)
  int counter, size, mode;
  Boolean orient, sticky;
{
  int i = mode - PERIOD2;
  int j = (orient) ? 1 : 0;
  int k = (sticky) ? MAXTETRAS - MINTETRAS + 1 : size - MINTETRAS;

  if (counter < pyraminx_record[i][j][k]) {
    pyraminx_record[i][j][k] = counter;
    if (size < 4 || mode == PERIOD2 || (orient &&
        (pyraminx_record[i][j][k] < pyraminx_record[i][!j][k])))
      pyraminx_record[i][!j][k] = counter;
    write_records();
    (void) sprintf(record_dsp, "%d", counter);
    return TRUE;
  }
  return FALSE;
}

static void print_state(w, prog, mode, size, sticky, moves, record, message)
  Widget w;
  char *prog, *record, *message;
  int mode, size, moves;
  Boolean sticky;
{
  char ss[10], mb[10];

  if (sticky)
    (void) strcpy(ss, "sticky");
  else
    (void) sprintf(ss, "%d", size);
  if (mode == BOTH)
    (void) strcpy(mb, "both");
  else
    (void) sprintf(mb, "%d", mode);
  (void) sprintf(title_dsp, "%s.%s: %s@ (%d/%s) - %s", prog, mb, ss, moves,
           record, message);
  XtSetArg(arg[0], XtNtitle, title_dsp);
  XtSetValues(w, arg, 1);
}

static void read_records()
{
  FILE *fp;
  int i, n, mode, orient;

  for (mode = 0; mode < MAXMODES; mode++)
    for (orient = 0; orient < 2; orient++)
      for (i = 0; i <= MAXTETRAS - MINTETRAS + 1; i++)
        pyraminx_record[mode][orient][i] = MAXRECORD;
  if ((fp = fopen(SCOREFILE, "r")) == NULL)
    (void) sprintf(message_dsp, "Can not open %s, taking defaults.", SCOREFILE);
  else {
    for (mode = 0; mode < MAXMODES; mode++)
      for (orient = 0; orient < 2; orient++)
        for (i = 0; i <= MAXTETRAS - MINTETRAS + 1; i++) {
          (void) fscanf(fp, "%d", &n);
          pyraminx_record[mode][orient][i] = n;
        }
    (void) fclose(fp);
  }
}

static void write_records()
{
  FILE *fp;
  int i, orient, mode;

  if ((fp = fopen(SCOREFILE, "w")) == NULL)
    (void) sprintf(message_dsp, "Can not write to %s.", SCOREFILE);
  else {
    for (mode = 0; mode < MAXMODES; mode++) {
      for (orient = 0; orient < 2; orient++) {
        for (i = 0; i <= MAXTETRAS - MINTETRAS + 1; i++)
          (void) fprintf(fp, "%d ", pyraminx_record[mode][orient][i]);
        (void) fprintf(fp, "\n");
      }
      (void) fprintf(fp, "\n");
    }
    (void) fclose(fp);
  }
}
